sheep/recovery.c

```diff
@@ -1151,6 +1158,111 @@ static void screen_object_list(struct recovery_list_work *rlw,
 	xqsort(rlw->oids, rlw->count, obj_cmp);
 }
 
+static int vnode_to_node_idx(struct sd_vnode *vnode, int nr_nodes,
+			     struct sd_node *nodes)
+{
+	for (int i = 0; i < nr_nodes; i++) {
+		if (node_id_cmp(&vnode->node->nid, &nodes[i].nid) == 0)
+			return i;
+	}
+
+	panic("vnode couldn't found in the node array");
+	return -1;		/* never executed */
+}
+
+struct seen_object_entry {
+	uint64_t oid;
+	struct rb_node node;
+};
+
+static int seen_object_cmp(const struct seen_object_entry *a,
+			   const struct seen_object_entry *b)
+{
+	return intcmp(a->oid, b->oid);
+}
+
+static bool check_diskfull_possibility(uint32_t epoch, struct vnode_info *vinfo,
+				       int nr_nodes, struct sd_node *nodes)
+{
+	uint64_t **oids_per_node;
+	size_t *nr_oids;
+	uint64_t *required_space_per_node;
+	const struct sd_vnode *vnodes[SD_MAX_COPIES];
+	bool ret = false;
+	struct rb_root seen_objects = RB_ROOT;
+
+	oids_per_node = xcalloc(nr_nodes, sizeof(uint64_t *));
+	nr_oids = xcalloc(nr_nodes, sizeof(size_t));
+	required_space_per_node = xcalloc(nr_nodes, sizeof(uint64_t));
+
+	for (int i = 0; i < nr_nodes; i++)
+		oids_per_node[i] = fetch_object_list(&nodes[i], epoch,
+						     &nr_oids[i]);
+
+	for (int i = 0; i < nr_nodes; i++) {
+		uint64_t *oids = oids_per_node[i];
+
+		for (int j = 0; j < nr_oids[i]; j++) {
+			int nr_objs = get_obj_copy_number(oids[j],
+							  vinfo->nr_zones);
+			struct seen_object_entry *key;
+
+			key = xzalloc(sizeof(*key));
+			key->oid = oids[j];
+
+			if (rb_search(&seen_objects, key, node,
+				      seen_object_cmp) != NULL) {
+				free(key);
+				continue;
+			}
+			rb_insert(&seen_objects, key, node, seen_object_cmp);
+
+			oid_to_vnodes(oids[j], &vinfo->vroot, nr_objs, vnodes);
+
+			for (int k = 0; k < nr_objs; k++) {
+				int node_idx = vnode_to_node_idx(
+					(struct sd_vnode *)vnodes[k],
+					nr_nodes, nodes);
+
+				/*
+				 * TODO: current calculation doesn't consider
+				 * about space consumption by metadata objects
+				 * e.g. inode, ledger
+				 */
```

Comment:
I think it's better to use ```get_objsize()``` to accumulate ```required_space_per_node[node_idx]``` when ```!is_data_obj(oids[j])``` is false.

https://github.com/sheepdog/sheepdog/blob/master/include/sheepdog_proto.h#L532